client/src/panels/panel_declarations/results/result_drilldown/result_drilldown.js

```diff
@@ -309,6 +311,38 @@ class SingleSubjExplorer extends React.Component {
                 onChange={(evt) => this.handleQueryChange(evt.target.value)}
                 value={query}
               />
+              <button
+                type="button"
+                className="btn btn-ib-light"
+                style={{
+                  height: "40px",
+                  width: "50%",
+                  margin: "5px 0px",
+                }}
+                onClick={() => {
+                  this.setState({ loading_query: true });
+                  expand_all(root);
+                  this.setState({ loading_query: false });
```

Comment:
First, if you're not familiar with the JS event loop architecture, I recommend googling about it and reading an article or two. Plenty of well written resources to explain it better than I can, and enough that I won't try to recommend a specific one. Just pick any that seem suited to your learning style and let me know if you have questions. It's not the direct culprit here, but it's an important player in all this. Going to be categorizing a lot of code as synchronous or asynchronous in both my explanation of what's gone wrong and in my suggestions for a fix.

Long story short, you're getting bit by a relatively new (as of React 16, the current major version number) feature, so it's definitely possible that myself or any of the tutorials you followed are to blame for glossing over this special case! You've probably been taught (by me?) that a `setState` immediately prompts a rerender... but there's currently ONE case where that's not true: inside of a React event handler!

State/prop ipdates occurring in the call stack of a React event handler are instead asynchronously batched for later, and if multiple prop/state changes for a single component fall in to the same batch queue than they're merged and applied as a single update. So, an event callback can only result in one rerender. Here, not only does the first `setState` end up clobbered by the second, but all the synchronous work done by `expand_all(root)` still needs to execute on the JS thread before `onClick` gets off the call stack and the asynchronous batched-render even has a chance to start!

For more nitty gritty on batched renders: https://stackoverflow.com/questions/48563650/does-react-keep-the-order-for-state-updates/48610973#48610973

The plan is, I believe, to eventually make ALL React rendering batch and occur asynchronously by default. Good for performance, but probably also a steeper learning curve for the whole framework... but it will be better than the current position, where it's super easy to be unaware of which `setState`s will synchronously re-render and which are going to be batched in to an asynchronous render.

So, options to subvert this "helpful" batching? 

One option: fight async with async
```
onClick={() => {
  this.setState({ loading_query: true });
  setTimeout(() => {
    expand_all(root);
    this.setState({ loading_query: false });
  }, 0);
}}
```
Moves `expand_all(root)` and `this.setState({ loading_query: false })` out of the `onClick` call stack and throws them on the JS message queue (even a timeout of 0 creates async code). Lets `onClick` resolve and the immediate async render for `this.setState({ loading_query: true })` occur before the the JS thread gets around to expanding nodes and removing the spinner. 

Complication 1: it's complicated. Requires all developers touching this code to understand the JS event loop and know what React's up to with batching renders caused by event handlers. There are enough timeouts flying around this file because of the debounced search input. If you don't dig up another solution, use this but add comments on the "why" of it.

Complication 2: need to do some book keeping to make sure timeouts clear when the component dies. See the `componentWillUnmount` lifecycle method, it's already tracking and killing timeouts related to the debounced search input. This is mostly because if `setState` is called on a dead component you get a lot of annoying console warnings about potential memory leaks, but it's also important because sometimes those warnings are right and there really is a memory leak.